internal/endpoints/util.go

```diff
@@ -0,0 +1,39 @@
+package endpoints
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net"
+	"net/http"
+	"time"
+
+	"github.com/canonical/lxd/shared/logger"
+)
+
+// shutdownServer will shutdown the given server.
+// If the given timeout is 0, it will forcefully shut it down. Otherwise, it will gracefully shut it down.
+func shutdownServer(ctx context.Context, server *http.Server, timeout time.Duration) error {
+	// If the given timeout is 0, force the shutdown.
+	if timeout == 0 {
+		err := server.Close()
+		if errors.Is(err, net.ErrClosed) {
+			return nil
+		}
+		return err
+	}
+
+	// server.Shutdown will gracefully stop the server, allowing existing requests to finish.
+	shutdownCtx, cancel := context.WithTimeout(ctx, timeout)
+	defer cancel()
+	err := server.Shutdown(shutdownCtx)
+	if err != nil {
+		logger.Error("Failed to gracefully shutdown server", logger.Ctx{"err": err})
+		if closeErr := server.Close(); closeErr != nil {
```

Comment:
Looks [this](https://github.com/canonical/microcluster/pull/295#discussion_r1879636542) was missed out.

Reply:
Done